wrapanapi/containers/providers/rhopenshift.py

```diff
@@ -76,6 +76,35 @@ def retrieve_cfme_appliance_version(template_name):
 """
 
 
+def reconnect(decorator):
+    def decorate(cls):
+        for attr in cls.__dict__:
+            if callable(getattr(cls, attr)) and not attr.startswith('_'):
+                setattr(cls, attr, decorator(getattr(cls, attr)))
+        return cls
+    return decorate
+
+
+def unauthenticated_error_handler(method):
+    """Fixes issue with 401 error by restoring connection.
+        Sometimes connection to openshift api endpoint gets expired and openshift returns 401.
+        As a result tasks in some applications like sprout fail.
+    """
+    @wraps(method)
+    def wrap(*args, **kwargs):
+        attempts = 3
+        for attempt in range(1, attempts + 1):
```

Comment:
Checking for "attempt <= attempts" is redundant since it's impossible for attempt to ever be > attempts inside your for loop. I think this could be written as:
```python
def wrap(*args, **kwargs):
    reconnect_attempts = 3
    for attempt in range(reconnect_attempts):
        try:
            return method(*args, **kwargs)
        except ApiException as e:
            if e.status == 401:
                args[0]._connect()
    # Final attempt, 401 exception won't be caught here if we still hit it.
    return method(*args, **kwargs)
```

That would give 4 total method attempts, with 3 reconnect attempts.

Reply:
I had the same code in the beginning but changed it to what you saw in order to make exact number of reconnect attempts. But it seems it isn't necessary :)